File: smart-agriculture-suite/src/FusionModule.cpp
#include "FusionModule.h"

#include <math.h>
#include <string.h>
#include <Preferences.h>

#include "AppConfig.h"

namespace agri {

namespace {
constexpr const char* kFusionNs = "fusion_nn";
}  // namespace

const char* FusionModule::decisionName(Decision decision) {
    switch (decision) {
        case Decision::Moderate: return "moderate";
        case Decision::Heavy: return "heavy";
        case Decision::None:
        default:
            return "none";
    }
}

void FusionModule::begin(bool autoControlEnabled) {
    autoControlEnabled_ = autoControlEnabled;
    initChannels();
    initNetwork();
    loadNetworkFromPrefs();
}

void FusionModule::update(const SensorSnapshot& snapshot, bool sampleUpdated, unsigned long nowMs, ActuatorController& actuator) {
    if (sampleUpdated) {
        const float rawValues[kSensorCount] = {
            snapshot.airTemp,
            snapshot.airHumi,
            snapshot.soilHumi,
            snapshot.lightValue,
            snapshot.liquidLevel
        };

        for (int i = 0; i < kSensorCount; ++i) {
            channels_[i].rawValue = rawValues[i];
            applyKalmanFilter(channels_[i], rawValues[i]);
            normalizeValue(channels_[i]);
            updateReliability(channels_[i]);
        }
        calculateWeights();
    }

    if (nowMs - lastFusionMs_ < kFusionIntervalMs) {
        return;
    }

    result_ = performFusion();
    lastFusionMs_ = nowMs;

    if (!autoControlEnabled_ || !actuator.status().autoMode || actuator.isBusy(nowMs)) {
        return;
    }

    unsigned long duration = 0;
    switch (result_.decision) {
        case Decision::Moderate: duration = 45UL; break;
        case Decision::Heavy: duration = 120UL; break;
        case Decision::None:
        default:
            duration = 0;
            break;
    }
    if (duration > 0) {
        actuator.startTimedRun(ControlSource::Fusion, duration, nowMs);
    }
}

void FusionModule::writeStatus(JsonDocument& doc) const {
    doc["autoControlEnabled"] = autoControlEnabled_;
    doc["decision"] = static_cast<int>(result_.decision);
    doc["decisionName"] = decisionName(result_.decision);
    doc["confidence"] = result_.confidence;
    doc["needScore"] = result_.needScore;
    doc["weightedScore"] = result_.weightedScore;
    doc["nnScore"] = result_.nnScore;
    doc["finalScore"] = result_.finalScore;
    doc["totalDecisions"] = totalDecisions_;
    doc["irrigationCount"] = irrigationCount_;
    doc["avgConfidence"] = averageConfidence_;

    doc["nn"]["none"] = output_[0];
    doc["nn"]["moderate"] = output_[1];
    doc["nn"]["heavy"] = output_[2];
}

void FusionModule::writeSensors(JsonDocument& doc) const {
    JsonArray sensors = doc.to<JsonArray>();
    for (int i = 0; i < kSensorCount; ++i) {
        JsonObject item = sensors.add<JsonObject>();
        item["name"] = channels_[i].name;
        item["label"] = channels_[i].label;
        item["unit"] = channels_[i].unit;
        item["raw"] = channels_[i].rawValue;
        item["filtered"] = channels_[i].kalmanEstimate;
        item["normalized"] = channels_[i].normalizedValue;
        item["reliability"] = channels_[i].reliability;
        item["weight"] = channels_[i].weight;
        item["kalmanGain"] = channels_[i].kalmanGain;
        item["healthy"] = channels_[i].healthy;
    }
}

void FusionModule::initChannels() {
    channels_[0].name = "AirTemp";
    channels_[0].label = "Temperature";
    channels_[0].unit = "C";
    channels_[0].kalmanEstimate = 25.0f;
    channels_[0].weight = 0.20f;
    channels_[0].maxRange = 40.0f;

    channels_[1].name = "AirHumi";
    channels_[1].label = "Humidity";
    channels_[1].unit = "%";
    channels_[1].kalmanEstimate = 60.0f;
    channels_[1].weight = 0.20f;
    channels_[1].maxRange = 100.0f;

    channels_[2].name = "SoilHumi";
    channels_[2].label = "Soil";
    channels_[2].unit = "%";
    channels_[2].kalmanEstimate = 50.0f;
    channels_[2].weight = 0.30f;
    channels_[2].maxRange = 100.0f;

    channels_[3].name = "Light";
    channels_[3].label = "Light";
    channels_[3].unit = "lux";
    channels_[3].kalmanEstimate = 500.0f;
    channels_[3].weight = 0.15f;
    channels_[3].maxRange = 10000.0f;

    channels_[4].name = "Liquid";
    channels_[4].label = "Liquid";
    channels_[4].unit = "%";
    channels_[4].kalmanEstimate = 80.0f;
    channels_[4].weight = 0.15f;
    channels_[4].maxRange = 100.0f;
}

void FusionModule::initNetwork() {
    // Trained weights from synthetic data (val acc 86.4%)
    const float presetIH[kSensorCount][kHidden] = {
        {-0.7618f, -0.1648f, 0.9222f, 0.2010f, 0.1822f, 0.8395f, 0.6211f, -0.7773f},
        {0.7061f, -0.4602f, -0.9656f, -0.1033f, -0.5191f, 0.2046f, -0.2469f, 1.1025f},
        {1.1686f, 0.1717f, -0.5193f, -0.1273f, -0.0344f, -0.5801f, -0.5640f, 1.7332f},
        {0.1232f, -0.2913f, 1.0482f, 0.3298f, -0.1933f, 0.3476f, 0.3208f, -0.1836f},
        {-0.4105f, 0.1005f, 1.5418f, -2.6921f, -0.4168f, 1.2474f, -2.1958f, -0.7266f},
    };
    const float presetBH[kHidden] = {
        0.5547f, -0.0495f, 0.2924f, 0.5884f, -0.0884f, 0.3348f, 0.4393f, 0.7546f
    };
    const float presetHO[kHidden][kOutputs] = {
        {1.1199f, 0.6752f, -2.4982f},
        {0.6507f, -0.6800f, 0.1269f},
        {-1.5823f, -0.1323f, 1.2406f},
        {2.7740f, -0.9054f, -1.8639f},
        {-0.6520f, 0.0906f, -0.5720f},
        {-1.1279f, 0.6628f, 0.5219f},
        {2.5613f, -2.2228f, -0.1253f},
        {1.4902f, -0.1788f, -3.4092f},
    };
    const float presetBO[kOutputs] = {
        -0.1035f, 0.3601f, -0.2272f
    };

    memcpy(weightsIH_, presetIH, sizeof(presetIH));
    memcpy(biasH_, presetBH, sizeof(presetBH));
    memcpy(weightsHO_, presetHO, sizeof(presetHO));
    memcpy(biasO_, presetBO, sizeof(presetBO));
}

void FusionModule::applyKalmanFilter(SensorChannel& channel, float measurement) {
    constexpr float kQ = 0.01f;
    constexpr float kR = 0.1f;

    const float predictedEstimate = channel.kalmanEstimate;
    const float predictedError = channel.kalmanError + kQ;

    channel.kalmanGain = predictedError / (predictedError + kR);
    channel.kalmanEstimate = predictedEstimate + channel.kalmanGain * (measurement - predictedEstimate);
    channel.kalmanError = (1.0f - channel.kalmanGain) * predictedError;
}

void FusionModule::normalizeValue(SensorChannel& channel) {
    const float range = channel.maxRange - channel.minRange;
    if (range <= 0.0f) {
        channel.normalizedValue = 0.0f;
        return;
    }

    channel.normalizedValue = (channel.kalmanEstimate - channel.minRange) / range;
    channel.normalizedValue = constrain(channel.normalizedValue, 0.0f, 1.0f);
}

void FusionModule::updateReliability(SensorChannel& channel) {
    float reliability = 1.0f;

    if (channel.rawValue <= 0.001f || channel.rawValue > channel.maxRange * 1.5f) {
        reliability *= 0.1f;
        ++channel.faultCount;
    } else {
        channel.faultCount = max(0, channel.faultCount - 1);
    }

    if (channel.kalmanGain > 0.8f) {
        reliability *= 0.7f;
    }

    if (channel.faultCount > 5) {
        reliability *= 0.3f;
        channel.healthy = false;
    } else {
        channel.healthy = true;
    }

    channel.reliability = channel.reliability * 0.8f + reliability * 0.2f;
}

void FusionModule::calculateWeights() {
    float totalReliability = 0.0f;
    for (int i = 0; i < kSensorCount; ++i) {
        totalReliability += channels_[i].reliability;
    }
    if (totalReliability <= 0.0f) {
        return;
    }
    for (int i = 0; i < kSensorCount; ++i) {
        channels_[i].weight = channels_[i].reliability / totalReliability;
    }
}

void FusionModule::runNeuralNetwork(const float inputs[], float outputs[]) {
    for (int h = 0; h < kHidden; ++h) {
        float sum = biasH_[h];
        for (int i = 0; i < kSensorCount; ++i) {
            sum += inputs[i] * weightsIH_[i][h];
        }
        hiddenOutput_[h] = relu(sum);
    }

    float raw[kOutputs];
    for (int o = 0; o < kOutputs; ++o) {
        float sum = biasO_[o];
        for (int h = 0; h < kHidden; ++h) {
            sum += hiddenOutput_[h] * weightsHO_[h][o];
        }
        raw[o] = sum;
    }

    softmax(raw, outputs, kOutputs);
    memcpy(output_, outputs, sizeof(output_));
}

float FusionModule::relu(float value) {
    return value > 0.0f ? value : 0.0f;
}

void FusionModule::softmax(float input[], float output[], int size) {
    float maxValue = input[0];
    for (int i = 1; i < size; ++i) {
        maxValue = max(maxValue, input[i]);
    }

    float sum = 0.0f;
    for (int i = 0; i < size; ++i) {
        output[i] = expf(input[i] - maxValue);
        sum += output[i];
    }
    for (int i = 0; i < size; ++i) {
        output[i] /= sum;
    }
}

FusionModule::FusionResult FusionModule::performFusion() {
    FusionResult result;

    float needFactors[kSensorCount];
    needFactors[0] = channels_[0].normalizedValue;
    needFactors[1] = 1.0f - channels_[1].normalizedValue;
    needFactors[2] = 1.0f - channels_[2].normalizedValue;
    needFactors[3] = channels_[3].normalizedValue * 0.5f;
    needFactors[4] = channels_[4].normalizedValue;

    result.weightedScore = 0.0f;
    for (int i = 0; i < kSensorCount; ++i) {
        result.weightedScore += needFactors[i] * channels_[i].weight;
    }
    result.weightedScore *= 100.0f;

    float inputs[kSensorCount];
    for (int i = 0; i < kSensorCount; ++i) {
        inputs[i] = channels_[i].normalizedValue;
    }

    float outputs[kOutputs];
    runNeuralNetwork(inputs, outputs);
    result.nnScore = outputs[1] * 50.0f + outputs[2] * 100.0f;
    result.finalScore = result.weightedScore * 0.6f + result.nnScore * 0.4f;
    result.needScore = result.finalScore;

    if (channels_[4].rawValue < 20.0f) {
        result.decision = Decision::None;
        result.confidence = 0.95f;
    } else if (result.finalScore > 65.0f) {
        result.decision = Decision::Heavy;
        result.confidence = min(1.0f, result.finalScore / 100.0f);
    } else if (result.finalScore > 35.0f) {
        result.decision = Decision::Moderate;
        result.confidence = 0.5f + (result.finalScore - 35.0f) / 60.0f;
    } else {
        result.decision = Decision::None;
        result.confidence = 1.0f - result.finalScore / 70.0f;
    }

    ++totalDecisions_;
    if (result.decision != Decision::None) {
        ++irrigationCount_;
    }
    averageConfidence_ = averageConfidence_ * 0.95f + result.confidence * 0.05f;
    return result;
}

void FusionModule::loadNetworkFromPrefs() {
    Preferences prefs;
    if (!prefs.begin(kFusionNs, true)) {
        return;
    }

    // Check if persisted weights exist (key "ver" acts as marker)
    if (prefs.getUInt("ver", 0) == 0) {
        prefs.end();
        return;
    }

    // weightsIH_: 5×8 = 40 floats stored as "wi<row><col>" compact keys
    for (int i = 0; i < kSensorCount; ++i) {
        char key[8];
        for (int h = 0; h < kHidden; ++h) {
            snprintf(key, sizeof(key), "w%d%d", i, h);
            weightsIH_[i][h] = prefs.getFloat(key, weightsIH_[i][h]);
        }
    }

    // biasH_: 8 floats
    for (int h = 0; h < kHidden; ++h) {
        char key[8];
        snprintf(key, sizeof(key), "bh%d", h);
        biasH_[h] = prefs.getFloat(key, biasH_[h]);
    }

    // weightsHO_: 8×3 = 24 floats
    for (int h = 0; h < kHidden; ++h) {
        char key[8];
        for (int o = 0; o < kOutputs; ++o) {
            snprintf(key, sizeof(key), "v%d%d", h, o);
            weightsHO_[h][o] = prefs.getFloat(key, weightsHO_[h][o]);
        }
    }

    // biasO_: 3 floats
    for (int o = 0; o < kOutputs; ++o) {
        char key[8];
        snprintf(key, sizeof(key), "bo%d", o);
        biasO_[o] = prefs.getFloat(key, biasO_[o]);
    }

    prefs.end();
    Serial.println("[Fusion] loaded trained weights from NVS");
}

void FusionModule::saveNetworkToPrefs() {
    Preferences prefs;
    if (!prefs.begin(kFusionNs, false)) {
        Serial.println("[Fusion] failed to open NVS for writing");
        return;
    }

    prefs.putUInt("ver", 1);  // marker that weights exist

    for (int i = 0; i < kSensorCount; ++i) {
        char key[8];
        for (int h = 0; h < kHidden; ++h) {
            snprintf(key, sizeof(key), "w%d%d", i, h);
            prefs.putFloat(key, weightsIH_[i][h]);
        }
    }

    for (int h = 0; h < kHidden; ++h) {
        char key[8];
        snprintf(key, sizeof(key), "bh%d", h);
        prefs.putFloat(key, biasH_[h]);
    }

    for (int h = 0; h < kHidden; ++h) {
        char key[8];
        for (int o = 0; o < kOutputs; ++o) {
            snprintf(key, sizeof(key), "v%d%d", h, o);
            prefs.putFloat(key, weightsHO_[h][o]);
        }
    }

    for (int o = 0; o < kOutputs; ++o) {
        char key[8];
        snprintf(key, sizeof(key), "bo%d", o);
        prefs.putFloat(key, biasO_[o]);
    }

    prefs.end();
    Serial.println("[Fusion] saved trained weights to NVS");
}

}  // namespace agri

File: smart-agriculture-atlas200dk/src/ai/fusion_module.py
"""
智润智慧农业套件 - Atlas 200I DK A2 版
传感器融合模块 - 5通道卡尔曼滤波 + 5->8->3神经网络混合决策

对应原ESP32项目的 FusionModule.h/FusionModule.cpp
使用NumPy加速矩阵运算, 算法逻辑完全一致
"""

import math
import json
import time
import logging
from dataclasses import dataclass
from typing import Optional, List

import numpy as np

from config.app_types import SensorSnapshot, ControlSource
from config.hardware_config import DATA_DIR

logger = logging.getLogger(__name__)


@dataclass
class SensorChannel:
    name: str = ""
    label: str = ""
    unit: str = ""
    raw_value: float = 0.0
    kalman_estimate: float = 0.0
    kalman_error: float = 1.0
    kalman_gain: float = 0.0
    normalized_value: float = 0.0
    reliability: float = 1.0
    weight: float = 0.2
    fault_count: int = 0
    healthy: bool = True
    min_range: float = 0.0
    max_range: float = 100.0


@dataclass
class FusionResult:
    decision: str = "none"  # "none", "moderate", "heavy"
    confidence: float = 0.0
    need_score: float = 0.0
    weighted_score: float = 0.0
    nn_score: float = 0.0
    final_score: float = 0.0


SENSOR_COUNT = 5
HIDDEN_SIZE = 8
OUTPUT_SIZE = 3

# 预训练权重 (与ESP32版一致的合成数据训练结果)
PRESET_WEIGHTS_IH = np.array([
    [-0.7618, -0.1648,  0.9222,  0.2010,  0.1822,  0.8395,  0.6211, -0.7773],
    [ 0.7061, -0.4602, -0.9656, -0.1033, -0.5191,  0.2046, -0.2469,  1.1025],
    [ 1.1686,  0.1717, -0.5193, -0.1273, -0.0344, -0.5801, -0.5640,  1.7332],
    [ 0.1232, -0.2913,  1.0482,  0.3295, -0.1933,  0.3476,  0.3208, -0.1836],
    [-0.4105,  0.1005,  1.5418, -2.6921, -0.4168,  1.2474, -2.1958, -0.7266],
], dtype=np.float32)

PRESET_BIAS_H = np.array([0.5547, -0.0495, 0.2924, 0.5884, -0.0884, 0.3348, 0.4393, 0.7546],
                           dtype=np.float32)

PRESET_WEIGHTS_HO = np.array([
    [ 1.1199,  0.6752, -2.4982],
    [ 0.6507, -0.6800,  0.1269],
    [-1.5823, -0.1323,  1.2406],
    [ 2.7740, -0.9054, -1.8639],
    [-0.6520,  0.0906, -0.5720],
    [-1.1279,  0.6628,  0.5211],
    [ 2.5613, -2.2228, -0.1250],
    [ 1.4902, -0.1788, -3.4092],
], dtype=np.float32)

PRESET_BIAS_O = np.array([-0.1035, 0.3601, -0.2272], dtype=np.float32)


class FusionModule:
    """5通道卡尔曼滤波 + 神经网络混合传感器融合"""

    FUSION_INTERVAL = 10.0  # 秒
    WEIGHTED_RATIO = 0.6
    NN_RATIO = 0.4

    def __init__(self, data_dir: str = DATA_DIR):
        self._data_dir = data_dir
        self._auto_control_enabled = True
        self._channels: List[SensorChannel] = []
        self._init_channels()

        self._weights_ih = PRESET_WEIGHTS_IH.copy()
        self._bias_h = PRESET_BIAS_H.copy()
        self._weights_ho = PRESET_WEIGHTS_HO.copy()
        self._bias_o = PRESET_BIAS_O.copy()

        self._result = FusionResult()
        self._nn_output = np.zeros(OUTPUT_SIZE)
        self._last_fusion_time = 0.0
        self._total_decisions = 0
        self._irrigation_count = 0
        self._average_confidence = 0.0

    def _init_channels(self):
        configs = [
            ("AirTemp", "Temperature", "C", 25.0, 0.20, 0.0, 40.0),
            ("AirHumi", "Humidity", "%", 60.0, 0.20, 0.0, 100.0),
            ("SoilHumi", "Soil", "%", 50.0, 0.30, 0.0, 100.0),
            ("Light", "Light", "lux", 500.0, 0.15, 0.0, 10000.0),
            ("Liquid", "Liquid", "%", 80.0, 0.15, 0.0, 100.0),
        ]
        self._channels = [
            SensorChannel(name=n, label=l, unit=u, kalman_estimate=e,
                          weight=w, min_range=mn, max_range=mx)
            for n, l, u, e, w, mn, mx in configs
        ]

    def begin(self, auto_control_enabled: bool = True):
        self._auto_control_enabled = auto_control_enabled
        self._load_network()

    def update(self, snapshot: SensorSnapshot, sample_updated: bool,
               now: float, actuator) -> Optional[FusionResult]:
        if sample_updated:
            raw_values = [
                snapshot.air_temp, snapshot.air_humi, snapshot.soil_humi,
                snapshot.light_intensity, snapshot.liquid_level
            ]
            for i, val in enumerate(raw_values):
                self._channels[i].raw_value = val
                self._apply_kalman_filter(self._channels[i], val)
                self._normalize_value(self._channels[i])
                self._update_reliability(self._channels[i])
            self._calculate_weights()

        if now - self._last_fusion_time < self.FUSION_INTERVAL:
            return None

        self._result = self._perform_fusion()
        self._last_fusion_time = now

        if (not self._auto_control_enabled or actuator.is_busy(now)):
            return self._result

        duration = 0.0
        if self._result.decision == "moderate":
            duration = 45.0
        elif self._result.decision == "heavy":
            duration = 120.0

        if duration > 0:
            actuator.start_timed_run(ControlSource.TimedRun, duration, now)

        return self._result

    # ------------------------------------------------------------------
    # 卡尔曼滤波
    # ------------------------------------------------------------------
    @staticmethod
    def _apply_kalman_filter(channel: SensorChannel, measurement: float):
        Q = 0.01
        R = 0.1
        predicted_estimate = channel.kalman_estimate
        predicted_error = channel.kalman_error + Q
        channel.kalman_gain = predicted_error / (predicted_error + R)
        channel.kalman_estimate = predicted_estimate + channel.kalman_gain * (measurement - predicted_estimate)
        channel.kalman_error = (1.0 - channel.kalman_gain) * predicted_error

    @staticmethod
    def _normalize_value(channel: SensorChannel):
        rng = channel.max_range - channel.min_range
        if rng <= 0:
            channel.normalized_value = 0.0
            return
        channel.normalized_value = max(0.0, min(1.0,
            (channel.kalman_estimate - channel.min_range) / rng))

    @staticmethod
    def _update_reliability(channel: SensorChannel):
        reliability = 1.0
        if channel.raw_value <= 0.001 or channel.raw_value > channel.max_range * 1.5:
            reliability *= 0.1
            channel.fault_count += 1
        else:
            channel.fault_count = max(0, channel.fault_count - 1)

        if channel.kalman_gain > 0.8:
            reliability *= 0.7

        if channel.fault_count > 5:
            reliability *= 0.3
            channel.healthy = False
        else:
            channel.healthy = True

        channel.reliability = channel.reliability * 0.8 + reliability * 0.2

    def _calculate_weights(self):
        total = sum(c.reliability for c in self._channels)
        if total <= 0:
            return
        for c in self._channels:
            c.weight = c.reliability / total

    # ------------------------------------------------------------------
    # 神经网络
    # ------------------------------------------------------------------
    def _run_neural_network(self, inputs: np.ndarray) -> np.ndarray:
        hidden = self._bias_h + inputs @ self._weights_ih
        hidden = np.maximum(hidden, 0.0)  # ReLU

        raw = self._bias_o + hidden @ self._weights_ho
        # Softmax
        exp_vals = np.exp(raw - np.max(raw))
        output = exp_vals / exp_vals.sum()
        self._nn_output = output
        return output

    # ------------------------------------------------------------------
    # 融合决策
    # ------------------------------------------------------------------
    def _perform_fusion(self) -> FusionResult:
        result = FusionResult()

        # 需求因子
        need_factors = [
            self._channels[0].normalized_value,                    # 温度高 -> 需水
            1.0 - self._channels[1].normalized_value,              # 湿度低 -> 需水
            1.0 - self._channels[2].normalized_value,              # 土壤干 -> 需水
            self._channels[3].normalized_value * 0.5,              # 光照强 -> 微增
            self._channels[4].normalized_value,                    # 液位高 -> 安全
        ]

        result.weighted_score = sum(
            f * c.weight for f, c in zip(need_factors, self._channels)
        ) * 100.0

        inputs = np.array([c.normalized_value for c in self._channels], dtype=np.float32)
        outputs = self._run_neural_network(inputs)
        result.nn_score = outputs[1] * 50.0 + outputs[2] * 100.0
        result.final_score = result.weighted_score * self.WEIGHTED_RATIO + result.nn_score * self.NN_RATIO
        result.need_score = result.final_score

        if self._channels[4].raw_value < 20.0:
            result.decision = "none"
            result.confidence = 0.95
        elif result.final_score > 65.0:
            result.decision = "heavy"
            result.confidence = min(1.0, result.final_score / 100.0)
        elif result.final_score > 35.0:
            result.decision = "moderate"
            result.confidence = 0.5 + (result.final_score - 35.0) / 60.0
        else:
            result.decision = "none"
            result.confidence = 1.0 - result.final_score / 70.0

        self._total_decisions += 1
        if result.decision != "none":
            self._irrigation_count += 1
        self._average_confidence = self._average_confidence * 0.95 + result.confidence * 0.05

        return result

    # ------------------------------------------------------------------
    # 权重持久化
    # ------------------------------------------------------------------
    def _load_network(self):
        import os
        path = os.path.join(self._data_dir, "fusion_weights.json")
        if os.path.exists(path):
            try:
                with open(path, 'r') as f:
                    data = json.load(f)
                self._weights_ih = np.array(data["weights_ih"], dtype=np.float32)
                self._bias_h = np.array(data["bias_h"], dtype=np.float32)
                self._weights_ho = np.array(data["weights_ho"], dtype=np.float32)
                self._bias_o = np.array(data["bias_o"], dtype=np.float32)
                logger.info("融合网络权重已加载")
            except Exception as e:
                logger.warning(f"融合网络权重加载失败: {e}")

    def save_network(self):
        import os
        os.makedirs(self._data_dir, exist_ok=True)
        path = os.path.join(self._data_dir, "fusion_weights.json")
        try:
            with open(path, 'w') as f:
                json.dump({
                    "weights_ih": self._weights_ih.tolist(),
                    "bias_h": self._bias_h.tolist(),
                    "weights_ho": self._weights_ho.tolist(),
                    "bias_o": self._bias_o.tolist(),
                }, f)
            logger.info("融合网络权重已保存")
        except Exception as e:
            logger.warning(f"融合网络权重保存失败: {e}")

    def to_dict(self) -> dict:
        return {
            "autoControlEnabled": self._auto_control_enabled,
            "decision": self._result.decision,
            "decisionName": self._result.decision,
            "confidence": round(self._result.confidence, 3),
            "needScore": round(self._result.need_score, 2),
            "weightedScore": round(self._result.weighted_score, 2),
            "nnScore": round(self._result.nn_score, 2),
            "finalScore": round(self._result.final_score, 2),
            "totalDecisions": self._total_decisions,
            "irrigationCount": self._irrigation_count,
            "avgConfidence": round(self._average_confidence, 3),
            "nn": {
                "none": round(float(self._nn_output[0]), 4),
                "moderate": round(float(self._nn_output[1]), 4),
                "heavy": round(float(self._nn_output[2]), 4),
            },
            "sensors": [
                {
                    "name": c.name, "label": c.label, "unit": c.unit,
                    "raw": round(c.raw_value, 2),
                    "filtered": round(c.kalman_estimate, 2),
                    "normalized": round(c.normalized_value, 4),
                    "reliability": round(c.reliability, 3),
                    "weight": round(c.weight, 4),
                    "kalmanGain": round(c.kalman_gain, 4),
                    "healthy": c.healthy,
                }
                for c in self._channels
            ],
        }
